File: src/ArrayOperations/54_SpiralMatrix/SpiralMatrix.cs
// 方向变量模拟路径 o(n)

public class Solution {
    public IList<int> SpiralOrder(int[][] matrix) 
    {
        IList<int> result = new List<int>();
        if (matrix.Length == 0 || matrix[0].Length == 0)
            return result;
        
        int m = matrix.Length;
        int n = matrix[0].Length;
        var visited = new bool[m, n];
        
        int i = 0;
        int j = 0;
        int direction = 1;
        
        while (true)
        {
            if (i < 0 || j < 0 || i == m || j == n || visited[i, j])
                break;
                 
            result.Add(matrix[i][j]);
            visited[i, j] = true;
            switch (direction)
            {
                case 1:
                    if (j + 1 == n || visited[i, j + 1]){
                        i++;
                        direction = 2;
                    } 
                    else {
                        j++;
                    }
                    break;
                case 2:
                    if (i + 1 == m || visited[i + 1, j]){
                        j--;
                        direction = 3;
                    } else{
                        i++;
                    }
                    break;
                case 3:
                    if (i == 0 || visited[i, j - 1]){
                        i--;
                        direction = 4;
                    } else{
                        j--;
                    }
                    break;
                case 4:
                    if (j == 0 || visited[i - 1, j]){
                        j++;
                        direction = 4;
                    } else{
                        i--;
                    }
                    break;
                default:
                    break;
            }
        }
        return result;
    }
}

File: src/ArrayOperations/54_SpiralMatrix/SpiralMatrix.py
# 方向变量模拟路径 o(n)

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        if (len(matrix) * len(matrix[0]) == 0):
            return []
        
        top = 0
        left = 0
        bottom = len(matrix) - 1
        right = len(matrix[0]) - 1 
        size = len(matrix) * len(matrix[0])
        
        out = []
        
        
        while(len(out) < size):
            
            for i in range(left, right+ 1):
                if (len(out) < size):
                    out.append(matrix[top][i])
            top += 1
            
            
            for i in range(top, bottom +1):
                if (len(out) < size):
                    out.append(matrix[i][right])
                
            right -= 1
            
            
            for i in range(right, left-1, -1):
                if (len(out) < size):
                    out.append(matrix[bottom][i])
                    
            bottom -= 1
            
            for i in range(bottom, top-1, -1):
                if (len(out) < size):
                    out.append(matrix[i][left])
            
            left += 1
            
            
            print(out)
            
        return out